### PR TITLE
feat: support TypeScript 5.7

### LINKED ISSUE #10115: TypeScript 5.7 Support
https://devblogs.microsoft.com/typescript/announcing-typescript-5-7-beta/

Please read through our TypeScript version policy: [typescript-eslint.io/users/dependency-versions#typescript](https://typescript-eslint.io/users/dependency-versions#typescript)

This issue is just to track all of the new features and their implementation state in this project.
As with all releases, we will not necessarily to support all features until closer to the full release when everything the features are stabilised.

_Please be patient._

---

# 🔲 Support for --target es2024 and --lib es2024

We'll need to update the [`getLib` function](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/parser/src/parser.ts#L46-L77).

# 🔲 Search Ancestor Configuration Files for Project Ownership

Perhaps we should follow their logic and update our `project: true` logic?

# 🔲 lib.d.ts Updates

We will need to regenerate our types within scope-manager.

# Other changes with no impact to us

- Checks for Never-Initialized Variables
- Path Rewriting for Relative Paths
- Faster Project Ownership Checks in Editors for Composite Projects
- Validated JSON Imports in --module nodenext
- Support for V8 Compile Caching in Node.js
- TypedArrays Are Now Generic Over ArrayBufferLike
- Creating Index Signatures from Non-Literal Method Names in Classes
- More Implicit any Errors on Functions Returning null and undefined

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

- [x] Addresses an existing open issue: fixes #10115
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Updates lib and scope definitions.

Doesn't update `project: true` to match the new ancestor configuration file searching, as that should already be covered by `projectService: true`.

💖

### CODE REVIEW

**Review by bradzacher — APPROVED** [review]

File: packages/types/src/lib.ts [hunk 1]
```diff
@@ -71,13 +72,21 @@ type Lib =
   | 'es2022.intl'
   | 'es2022.object'
   | 'es2022.regexp'
-  | 'es2022.sharedmemory'
```
Comment (burtek) [inline packages/types/src/lib.ts:74]: Newest typescript might have removed this, but the plugin should be compatible with previous TS versions (as per range in package.json) and those previous versions do know what `es2022.sharedmemory` is. This change makes this eslint plugin incompatible with previous TS versions in projects declaring use of `es2022.sharedmemory`

Comment (JoshuaKGoldberg) [reply packages/types/src/lib.ts:74]: Interesting. Could you please file an issue with a reproduction?

https://typescript-eslint.io/contributing/pull-requests

Comment (burtek) [reply packages/types/src/lib.ts:74]: Sure, on it

Comment (burtek) [reply packages/types/src/lib.ts:74]: https://github.com/typescript-eslint/typescript-eslint/issues/10395

Comment (burtek) [reply packages/types/src/lib.ts:74]: Actually, it's not this change, it's the rename of the lib file. Still caused by this PR

**Pull request merged**